### PR TITLE
Migrate all managers under system package to kotlin

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Towards #532 

@JSunde PTAL?

### CODE REVIEW

File: gnd/src/main/java/com/google/android/gnd/system/LocationManager.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+     * up will only see the latest location.
+     */
+    fun getLocationUpdates(): Flowable<Location> {
+        // There sometimes noticeable latency between when location update request succeeds and when
+        // the first location update is received. Requesting the last know location is usually
+        // immediate, so we merge into the stream to reduce perceived latency.
+        return locationUpdates
+            .startWith(locationClient.lastLocation.toObservable())
+            .toFlowable(BackpressureStrategy.LATEST)
+    }
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/system/LocationManager.kt:67]: nit: could be single-line expression function

File: gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     * Returns a [Completable] that completes once the specified permission is granted or
+     * terminates with error [PermissionDeniedException] if the requested permission was denied.
+     */
+    private fun getPermissionsResult(permission: String): @Cold Completable {
+        return activityStreams
+            .getNextRequestPermissionsResult(PERMISSIONS_REQUEST_CODE)
+            .flatMapCompletable { r: RequestPermissionsResult ->
+                completeOrError({ r.isGranted(permission) }, PermissionDeniedException::class.java)
+            }
+    }
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt:84]: ditto

File: gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    private fun onCheckSettingsFailure(throwable: Throwable): Completable {
+        return if (throwable is ResolvableApiException) {
+            val requestCode = LOCATION_SETTINGS_REQUEST_CODE
+            startResolution(requestCode, throwable).andThen(getNextResult(requestCode))
+        } else {
+            Completable.error(throwable)
+        }
+    }
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt:66]: ditto

File: gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            activityStreams.withActivity {
+                try {
+                    resolvableException.startResolutionForResult(it, requestCode)
+                    emitter.onComplete()
+                } catch (e: SendIntentException) {
+                    emitter.onError(e)
+                }
+            }
+        }
+    }
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt:83]: ditto

File: gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    override fun signOut() {
+        firebaseAuth.signOut()
+        signInState.onNext(SignInState(SignInState.State.SIGNED_OUT))
+        activityStreams.withActivity { getGoogleSignInClient(it).signOut() }
+    }
+
+    private fun getGoogleSignInClient(activity: Activity): GoogleSignInClient {
+        // TODO: Use app context instead of activity?
+        return GoogleSignIn.getClient(activity, googleSignInOptions)
+    }
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt:103]: ditto

File: gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    private fun FirebaseUser.toUser(): User {
+        return User.builder()
+            .setId(uid)
+            .setEmail(email)
+            .setDisplayName(displayName)
+            .setPhotoUrl(photoUrl.toString())
+            .build()
+    }
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt:141]: ditto

File: gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/MapContainerViewModel.java [hunk 7]
```diff
@@ -338,13 +338,20 @@ private Flowable<CameraUpdate> createLocationLockCameraUpdateFlowable(BooleanOrE
...
         .concatWith(locationUpdates.map(CameraUpdate::pan).skip(1));
   }
 
+  private static Point toPoint(Location location) {
+    return Point.newBuilder()
+        .setLatitude(location.getLatitude())
+        .setLongitude(location.getLongitude())
+        .build();
+  }
+
```
Comment (JSunde) [inline gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/MapContainerViewModel.java:354]: ditto

**Review by JSunde — CHANGES_REQUESTED** [review]

File: gnd/src/main/java/com/google/android/gnd/system/LocationManager.kt (hunk 1 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/LocationManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/LocationManager.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    // TODO: Request/remove updates on resume/pause.
+    @Synchronized
+    fun disableLocationUpdates(): Single<BooleanOrError> {
+        // Ignore errors when removing location updates, usually caused by disabling the same callback
+        // multiple times.
+        return removeLocationUpdates()
+            .toSingle { falseValue() }
+            .doOnError { Timber.e(it, "disableLocationUpdates") }
+            .onErrorReturn { falseValue() }
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/LocationManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    private val activityStreams: ActivityStreams
+) {
+    /**
+     * Asynchronously requests the app be granted the specified permission. If the permission has
+     * already been granted, completes immediately, otherwise completes once the next permissions
+     * result is received.
+     */
+    fun obtainPermission(permission: String): @Cold Completable {
+        return completeIf { requestPermission(permission) }.ambWith(getPermissionsResult(permission))
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            activityStreams.withActivity { activity: Activity ->
+                ActivityCompat.requestPermissions(
+                    activity,
+                    arrayOf(permission),
+                    PERMISSIONS_REQUEST_CODE
+                )
+            }
+            false
+        }
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/NetworkManager.kt [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * Returns a Completable that completes immediately on subscribe if network is available, or fails
+     * in error if not.
+     */
+    @JvmStatic
+    fun requireActiveNetwork(context: Context): Completable {
+        return completeOrError({ isNetworkAvailable(context) }, ConnectException::class.java)
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/NetworkManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    /** Returns `true` iff the app has been granted the specified permission.  */
+    private fun isGranted(permission: String): Boolean {
+        return checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt (hunk 2 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/PermissionsManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt (hunk 3 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    private fun getNextResult(requestCode: Int): Completable {
+        return activityStreams
+            .getNextActivityResult(requestCode)
+            .flatMapCompletable {
+                completeOrError({ it.isOk() }, SettingsChangeRequestCanceled::class.java)
+            }
+            .doOnComplete { Timber.d("Settings change request successful") }
+            .doOnError { Timber.e(it, "Settings change request failed") }
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/SettingsManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+    private fun onFirebaseAuthSuccess(authResult: AuthResult) {
+        // TODO: Store/update user profile in Firestore.
+        // TODO: Store/update user profile and image locally.
+        signInState.onNext(SignInState(authResult.user!!.toUser()))
+    }
+
+    private fun getFirebaseAuthCredential(googleAccount: GoogleSignInAccount): AuthCredential {
+        return GoogleAuthProvider.getCredential(googleAccount.idToken, null)
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt (hunk 6 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt (hunk 5 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/system/auth/GoogleAuthenticationManager.kt]: Done

File: gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/MapContainerViewModel.java (hunk 7 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/MapContainerViewModel.java:354]: This is not kotlin. Ignoring.

**Review by JSunde — APPROVED** [review]

**Pull request merged**